diff --git a/clang/lib/Driver/ToolChains/BezOS.h b/clang/lib/Driver/ToolChains/BezOS.h
new file mode 100644
index 0000000..515a3c2
--- /dev/null
+++ b/clang/lib/Driver/ToolChains/BezOS.h
@@ -0,0 +1,27 @@
+//===--- BezOS.h - BezOS ToolChain Implementations ----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_DRIVER_TOOLCHAINS_BezOS_H
+#define LLVM_CLANG_LIB_DRIVER_TOOLCHAINS_BezOS_H
+
+#include "Gnu.h"
+#include "clang/Driver/ToolChain.h"
+
+namespace clang {
+namespace driver {
+namespace toolchains {
+
+std::unique_ptr<ToolChain> getBezOSToolChain(const Driver &D,
+                                             const llvm::Triple &Triple,
+                                             const llvm::opt::ArgList &Args);
+
+} // end namespace toolchains
+} // end namespace driver
+} // end namespace clang
+
+#endif // LLVM_CLANG_LIB_DRIVER_TOOLCHAINS_BezOS_H
